src/Elastic.Apm/Config/AbstractConfigurationReader.cs

```diff
@@ -345,6 +351,38 @@ protected string ParseServiceName(ConfigurationKeyValue kv)
 			return ConfigConsts.DefaultValues.UnknownServiceName;
 		}
 
+		private string DiscoverServiceVersion()
+		{
+			var entryAssembly = Assembly.GetEntryAssembly();
+			if (entryAssembly != null && !IsMsOrElastic(entryAssembly.GetName().GetPublicKeyToken()))
+				return entryAssembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>().InformationalVersion;
+
+			return null;
+		}
+
+		protected string ParseServiceVersion(ConfigurationKeyValue kv)
+		{
+			var versionInConfig = kv.Value;
+
+			if (!string.IsNullOrEmpty(versionInConfig))
+			{
+				return versionInConfig;
+			}
+
+			Logger?.Info()?.Log("The agent was started without a service version. The service version will be automatically discovered.");
+
+			var discoveredVersion = DiscoverServiceVersion();
+			if (discoveredVersion != null)
+			{
+				Logger?.Info()
+					?.Log("The agent was started without a service version. The automatically discovered service version is {ServiceVersion}",
+						discoveredVersion);
+				return discoveredVersion;
+			}
+
+			return string.Empty;
```

Comment:
2 small things to this:
- Could we also add an info log before we return here? Otherwise in the logs in case the `if` block above is not entered we'll just print the `...version will be automatically discovered.` text, but won't send any version, which could be confusing.
- Any reason we return `string.Empty` here? Other methods return `null`. The general concept (`null` vs `Empty`) can be of course discussed, but I think mixing them (so having some methods returning `null`, some `Empty`) is I think the least favorable option.

Reply:
fixed